### DISCUSSION TITLE
`toYaml` template function and indentation

### DISCUSSION THREAD

**Discussion by vbrand1984** [twpayne/chezmoi#3873] [General]
I've noticed that the `toYaml` function puts 4 spaces per indentation level in its resulting yaml document. Though it doesn't go against the yaml specification (and chezmoi's `toYaml` inserts indentations correctly), but the standard de-facto is 2 spaces per nested level, and I wonder if this may cause issues when combining manually generated yaml text with template-generated fragments (currently I'm tinkering with Ansible and trying to manage its configuration with chezmoi).

**Comment by bradenhilton** [comment]
We'd have to use [`Encoder.SetIndent`](https://pkg.go.dev/gopkg.in/yaml.v3#Encoder.SetIndent).

In cases where consistent formatting is important, it's probably better to just round-trip the file with `fromYaml | toYaml` where possible.

**Reply by halostatue** [reply]
I don't think that I’ve used `toYaml` heavily before, but it *might* be worth changing the default indentation to 2 instead of 4, as it *is* typical to use 2 spaces for indentation in YAML.

**Reply by twpayne** [reply]
Changing the default will break existing code which relies on the current behavior. Adding an extra optional argument to `toYaml` will break template code that uses pipelines.

I don't think there's any standard indent in YAML. I can't find any mention of any in [the specification](https://yaml.org/spec/). chezmoi uses the default indentation used by `gopkg.in/yaml.v3`.

I think the only solution is to add a new `toYamlWithIndent` template function that takes the indentation level as an argument.

Thoughts?

**Reply by halostatue** [reply]
No, there's no standard indent, but I think that most projects use two space indents. Adding `toYamlWithIndent` would *sort* of work with the corresponding sprig function `toJsonPretty` (or is it `toPrettyJson`).

**Reply by twpayne** [reply]
I've added `toYamlWithIndent` in #3875. Reviews appreciated.

**Reply by bradenhilton** [reply]
I think it's worth adding overall, but `yaml.v3` doesn't appear to offer a way to change the indentation for arrays etc., so there is still a potential for inconsistent formatting when working with only part of a file.

Both of these are valid YAML:

```yaml
foos:
- bar
- baz
```

```yaml
foos:
  - bar
  - baz
```

I still think the best catch-all solution is to round-trip the formatting.
